contracts/governance/extensions/GovernorTimelockCompound.sol

```diff
@@ -0,0 +1,251 @@
+    function execute(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory calldatas,
+        bytes32 descriptionHash
+    ) public payable virtual override(IGovernor, Governor) returns (uint256) {
+        uint256 proposalId = hashProposal(targets, values, calldatas, descriptionHash);
+        Address.sendValue(payable(_timelock), msg.value);
+
+        uint256 eta = proposalEta(proposalId);
+        require(eta > 0, "GovernorTimelockCompound: proposal not yet queued");
+        for (uint256 i = 0; i < targets.length; ++i) {
+            _timelock.executeTransaction(targets[i], values[i], "", calldatas[i], eta);
+        }
+        _proposalTimelocks[proposalId].executed = true;
+
+        emit ProposalExecuted(proposalId);
+
+        return proposalId;
+    }
+
+    /**
+     * @dev Overriden version of the {Governor-_cancel} function to cancel the timelocked proposal if it as already
+     * been queued.
+     */
+    function _cancel(
+        address[] memory targets,
+        uint256[] memory values,
+        bytes[] memory calldatas,
+        bytes32 descriptionHash
+    ) internal virtual override returns (uint256) {
+        uint256 proposalId = super._cancel(targets, values, calldatas, descriptionHash);
+
+        uint256 eta = proposalEta(proposalId);
+        if (eta > 0) {
+            for (uint256 i = 0; i < targets.length; ++i) {
+                _timelock.cancelTransaction(targets[i], values[i], "", calldatas[i], eta);
+            }
+            _proposalTimelocks[proposalId].timer.reset();
+        }
+
+        return proposalId;
+    }
+
+    /**
+     * @dev Address through which the governor executes action. In this case, the timelock.
+     */
+    function _executor() internal view virtual override returns (address) {
+        return timelock();
+    }
+
+    /**
+     * @dev Accept admin right over the timelock.
+     */
+    // solhint-disable-next-line private-vars-leading-underscore
+    function __acceptAdmin() public {
+        _timelock.acceptAdmin();
+    }
+
+    /**
+     * @dev Public endpoint to update the underlying timelock instance. Restricted to the timelock itself, so updates
+     * must be proposed, scheduled and executed using the {Governor} workflow.
+     *
+     * For security reason, the timelock must be handed over to another admin before setting up a new one. The two
+     * operations (hand over the timelock) and do the update can be batched in a single proposal.
+     */
+    function updateTimelock(address newTimelock) external virtual onlyGovernance {
+        require(
+            _timelock.pendingAdmin() != address(0),
+            "GovernorTimelockCompound: old timelock must be transfered before update"
+        );
+        _updateTimelock(newTimelock);
+    }
+
+    function _updateTimelock(address newTimelock) private {
+        emit TimelockChange(address(_timelock), newTimelock);
+        _timelock = ICompoundTimelock(payable(newTimelock));
```

Comment:
If `newTimelock` needs to be payable we can declare that directly in the argument type, here and in the constructor as well.

Reply:
I don't really like having payable instead of address. It create a discrepancy between the function declaration and its selector (still computed using address). Also, no payment is actually going to happen here, so I think it's better not to create confusion in the eye of whoever is calling the `updateTimelock` function.